Introduce experimental flag/option (#16229)
ethereum/solidity: docs/using-the-compiler.rst

        "stopAfter": "parsing",
        // Optional: List of remappings
        "remappings": [ ":g=/dir" ],
        // Optional: Experimental mode toggle (Default: false)

>         "experimental": false,

Reviewer: ```suggestion
        // Optional: Experimental mode toggle (Default: false)
        // Makes it possible to use experimental features (but does not enable any such feature by itself).
        // The use of this mode is recorded in contract metadata.
        "experimental": false,
```

          //     - `<end>` is the index of the first byte after that location.
          // - `snippet`: A single-line code snippet from the location indicated by `@src`.
          //     The snippet is quoted and follows the corresponding `@src` annotation.
          // - `ast-id`: Annotations of the form `@ast-id <id>` over elements that can be mapped back to a definition in the original Solidity file.
          //   `<id>` is a node ID in the Solidity AST ('ast' output).
          // - `ethdebug`: Ethdebug annotations (experimental).
          // - `*`: Wildcard value that can be used to request all non-experimental components.

>           "debugInfo": ["location", "snippet", "ast", "ethdebug"]

Reviewer: ```suggestion
          "debugInfo": ["location", "snippet", "ast-id", "ethdebug"]
```

          // - `snippet`: A single-line code snippet from the location indicated by `@src`.
          //     The snippet is quoted and follows the corresponding `@src` annotation.
          // - `*`: Wildcard value that can be used to request everything.
          // NOTE: ethdebug option is experimental

>           "debugInfo": ["location", "snippet", "ast", "ethdebug"]

Reviewer: ```suggestion
// - `ast-id`: Annotations of the form `@ast-id <id>` over elements that can be mapped back to a definition in the original Solidity file.
//   `<id>` is a node ID in the Solidity AST ('ast' output).
// - `ethdebug`: Ethdebug annotations (experimental).
// - `*`: Wildcard value that can be used to request all non-experimental components.
```

        //   userdoc - User documentation (natspec)
        //   metadata - Metadata
        //   ir - Yul intermediate representation of the code before optimization
        //   irAst - AST of Yul intermediate representation of the code before optimization (experimental)
        //   irOptimized - Intermediate representation after optimization
        //   irOptimizedAst - AST of intermediate representation after optimization (experimental)
        //   storageLayout - Slots, offsets and types of the contract's state variables in storage
        //   transientStorageLayout - Slots, offsets and types of the contract's state variables in transient storage
        //   evm.assembly - New assembly format
        //   evm.legacyAssembly - Old-style assembly format in JSON
        //   evm.bytecode.ethdebug - Ethdebug output of the contract, requires viaIR (experimental)

>         //   evm.deployedBytecode.ethdebug - Deployed Ethdebug output of the contract, requires viaIR (experimental)

Reviewer: ```suggestion
        //   evm.bytecode.ethdebug - Debug information in ethdebug format (ethdebug/format/program schema). Can only be requested when compiling via IR. (experimental)
        //   evm.deployedBytecode.ethdebug - Like evm.deployedBytecode.ethdebug, but for the runtime part of the contract (experimental)
```

        //   userdoc - User documentation (natspec)
        //   metadata - Metadata
        //   ir - Yul intermediate representation of the code before optimization
        //   irAst - AST of Yul intermediate representation of the code before optimization (experimental)
        //   irOptimized - Intermediate representation after optimization
        //   irOptimizedAst - AST of intermediate representation after optimization (experimental)
        //   storageLayout - Slots, offsets and types of the contract's state variables in storage
        //   transientStorageLayout - Slots, offsets and types of the contract's state variables in transient storage
        //   evm.assembly - New assembly format
        //   evm.legacyAssembly - Old-style assembly format in JSON
        //   evm.bytecode.ethdebug - Debug information in ethdebug format (ethdebug/format/program schema). Can only be requested when compiling via IR. (experimental)

>         //   evm.deployedBytecode.ethdebug - Like evm.deployedBytecode.ethdebug, but for the runtime part of the contract (experimental)

Reviewer: Looking at it now, I see that I used the wrong output name in my suggestion.

```suggestion
        //   evm.deployedBytecode.ethdebug - Like evm.bytecode.ethdebug, but for the runtime part of the contract (experimental)
```

Sorry for that. But you're not paying attention :P

        //   evm.deployedBytecode.immutableReferences - Map from AST ids to bytecode ranges that reference immutables
        //   evm.methodIdentifiers - The list of function hashes
        //   evm.gasEstimates - Function gas estimates
        //   yulCFGJson - Control Flow Graph (CFG) of the Single Static Assignment (SSA) form of the contract (experimental)
        //
        // Note that using `evm`, `evm.bytecode`, etc. will select every
        // target part of that output. Additionally, `*` can be used as a wildcard to request everything.

>         //g

Reviewer: ```suggestion
        //
```

              "legacyAssembly": {},
              // Bytecode and related details.
              "bytecode": {
                // Ethdebug output (experimental)

>                 "ethdebug": {/* ... */},

Reviewer: You should also show the global `ethdebug` output. I think we should remove it (#16460), but for now it's there.

13. ``YulException``: Error during Yul code generation - this should be reported as an issue.
14. ``Warning``: A warning, which didn't stop the compilation, but should be addressed if possible.
15. ``Info``: Information that the compiler thinks the user might find useful, but is not dangerous and does not necessarily need to be addressed.

.. index:: ! Experimental mode, ! --experimental
.. _experimental-mode:

Experimental Mode
*****************

> Experimental mode has been introduced as of version 0.8.34 as an additional safeguard when using experimental features.

Reviewer 2: ```suggestion
Experimental mode has been introduced in version 0.8.34 as an additional safeguard when using experimental features.
```

I was tempted to suggest `Version` over `version` because technically it's a name when you refer to a specific version and should be capitalized. But it's consistently lower-case throughout the documentation (at least from what I could tell by a few glances). Meh :D

Prior to version 0.8.34, all experimental features not requiring an explicit ``pragma experimental ...`` in source file(s)
were freely available to be used without any additional safeguards. Given that experimental features are sparsely documented,
if at all, often not adequately tested, and thus not intended for production use, we decided to add  the ``--experimental``
flag to the command line interface, and the analogous ``settings.experimental`` boolean setting to the Standard JSON interface.

> Going forward, usage of any experimental feature will require the experimental mode to be toggled.

Reviewer: This feels more like something that would go into the release notes. In docs we usually just focus on describing the current state rather than the future and the past.

It's also missing some information. Here's my attempt at documenting it:

```suggestion
Some language and compiler features included in stable releases are not themselves considered stable.
They are sparsely documented, if at all, often not adequately tested, and thus not yet intended for production use.
In many cases it is possible to develop a big feature incrementally, with each iteration being already stable.
Sometimes, however, it is preferable to start with a prototype and stabilize it over multiple releases, while receiving feedback from users.
To prevent accidental use, such features can be only accessed by enabling the experimental mode.

There are no backwards compatibility guarantees for experimental features.
They are subject to change in breaking ways in non-breaking releases of the compiler.
Only major changes affecting them are recorded in the changelog.

To enable the experimental mode, use the ``--experimental`` flag on the command line,
or the analogous ``settings.experimental`` boolean setting in the Standard JSON input.
Note that the use of this mode is recorded in the metadata:

- ``experimental`` flag in CBOR metadata is set to ``true``,
- ``settings.experimental`` in JSON metadata is set to ``true``,
- ``experimentalFeatures`` field in JSON metadata includes the IDs of all the experimental features which may have affected the bytecode.

.. note::
    Prior to version 0.8.34, most of the experimental features were usable without any extra safeguards.
    Some were gated behind ``pragma experimental``, but this was not done consistently.
    The information about them was also only recorded in CBOR metadata and even then not always.
    The main goal of the experimental mode is to systematize this and make users fully when they are relying on features which are unfinished or not production-ready.
```

I'd also include some links, e.g. to changelog, metadata, section on pragmas.

Reviewer: Actually, I don't think the PR adds the metadata info. Full extent of it could be done separately, but we need to at least record at the use of `--experimental` so that the bytecode can be reproduced based on from metadata.

Author: Ah I missed `experimentalFeatures` - need to implement that as well.

Reviewer: That bit is fine in a follow-up PR. We also have an annotation for this on `SourceUnit` already, but it only contains features from experimental pragmas. We'll have to reconcile these two mechanisms as well.

Reviewer: By the way, don't forget to mention anything you add to metadata in [metadata docs](https://docs.soliditylang.org/en/develop/metadata.html).

| Feature               | ID                       | Affects bytecode | Flag/pragma                                                       |
+=======================+==========================+==================+===================================================================+
| AST import            | ``ast-import``           | yes              | ``--import-ast``                                                  |
+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+
| LSP                   | ``lsp``                  | yes              | ``--lsp``                                                         |
+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+
| EVM Assembly import   | ``evmasm-import``        | yes              | ``--import-asm-json``                                             |
+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+

> | Generic Solidity      | ``generic-solidity``     | yes              | ``pragma experimental solidity``                                  |

Reviewer: ```suggestion
| Core Solidity         | ``core-solidity``        | yes              | ``pragma experimental solidity``                                  |
```

In the we went with a different name for Experimental Solidity, but the spec was written before that.

+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+
| IR AST                | ``ir-ast``               | no               | ``--ir-ast-json``, ``--ir-optimized-ast-json``                    |
+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+
| EOF                   | ``eof``                  | yes              | ``--eof-version``                                                 |
+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+

> | Osaka EVM             | ``evm-osaka``            | yes              | ``--evm-version osaka``                                           |

Author: Wait, fusaka fork was in December, no? So this should be removed. I also didn't implement an evm version check for >current.

Author: Hm, not only that, Osaka is our current version; how could I go about testing the implementation?

Reviewer: It wasn't yet released when I created the spec :) Now we have no experimental EVM versions, but you could add Amsterdam already if you want to have one for testing. It's supposed to be ready in half a year, so we can already start preparing.

Reviewer: Also, now that I think about it, we could change the ID here to just `evm` covering all non-mainnet EVMs. This way we won't have to be all that fine-grained.

+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+
| Ethdebug              | ``ethdebug``             | no               | ``--ethdebug``, ``--ethdebug-runtime``, ``--debug-info ethdebug`` |
+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+
| Yul SSA CFG exporter  | ``yul-ssa-cfg-exporter`` | no               | ``--yul-cfg-json``                                                |
+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+

> | SSA CFG codegen       | ``ssa-cfg-codegen``      | yes              | ``--ssa-cfg-codegen`` (not yet on ``develop``)                    |

Reviewer: @clonker Do we actually need two separate flags for this? Like, are we planning to have the exporter stable before the codegen? Or do we need people to be able to export the CFG without enabling the codegen? To me they seem like parts of the same bigger feature.

Reviewer: Also, the docs should be written from the perspective of a release. So if it's not merged into `develop`, it should not be listed in the docs. It should be added as a part of the PR that introduces the flag.

Reviewer 2: the exporter was developed for and is used by project grey. it has also been merged to develop. the codegen hasn't.

Reviewer: I'd like to still clarify this before we finalize the PR.

Is it really just the exporter that's experimental? Isn't it rather the underlying CFG? The way I see it, the export is just a trivial part of it (just a JSON dump and some glue code) and the core of unproven logic is actually the calculation and structure of the CFG itself. I'd rename this flag to `ssa-cfg` to reflect that. The exporter happens to be the only merged feature that uses it, but if we add another one later, it should not need a new experimental flag IMO. The experimentalness comes from the CFG, not the export.

Following that reasoning, the SSA-CFG codegen should set the `ssa-cfg` flag as well, because it uses the same experimental CFG the exporter does.

Now, the codegen is more than just the CFG so we could have an extra `ssa-cfg-codegen` to distinguish it from pure CFG use. I'm just not sure we need that distinction. What matters is whether any CFG-related functionality was used and whether it affected the bytecode. We achieve this already by having a `ssa-cfg` flag and by recording it in metadata or not (depending on whether SSA-CFG codegen was used for a particular contract).

If we have a separate flag for the codegen, it would also make sense to have one e.g. for the SSA-CFG optimizer and I just don't think we need this level of granularity.

Reviewer: I discussed it with @clonker and @r0qs on Matrix and we agreed that having one `ssa-cfg` metadata feature flag for both is fine. So let's change it in the table.

We also agreed to rename the `--ssa-cfg-codegen` command-line flag to `--via-ssa-cfg` and have it automatically trigger `--via-ir`. But it's not something we'll be changing in this PR. The docs here should reflect the current state.

13. ``YulException``: Error during Yul code generation - this should be reported as an issue.
14. ``Warning``: A warning, which didn't stop the compilation, but should be addressed if possible.
15. ``Info``: Information that the compiler thinks the user might find useful, but is not dangerous and does not necessarily need to be addressed.

.. index:: ! Experimental mode, ! --experimental
.. _experimental-mode:

Experimental Mode
*****************

> 
> Some language and compiler features included in stable releases are not themselves considered stable.
> They are sparsely documented, if at all, often not adequately tested, and thus not yet intended for production use.
> In many cases it is possible to develop a big feature incrementally, with each iteration being already stable.
> Sometimes, however, it is preferable to start with a prototype and stabilize it over multiple releases, while receiving feedback from users.
> To prevent accidental use, such features can be only accessed by enabling the experimental mode.
> There are no backwards compatibility guarantees for experimental features.
> They are subject to change in breaking ways in non-breaking releases of the compiler.
> Only major changes affecting them are recorded in the changelog.
> To enable the experimental mode, use the ``--experimental`` flag on the command line,
> or the analogous ``settings.experimental`` boolean setting in the Standard JSON input.
> Note that the use of this mode is recorded in the metadata:
> - ``experimental`` flag in CBOR metadata is set to ``true``,
> - ``settings.experimental`` in JSON metadata is set to ``true``,

Reviewer: Please add some paragraph breaks. Otherwise it all gets bunched up into one big lump of text.

Also, you need an empty line before lists in RST (see how it looks without it in preview).

```suggestion
Sometimes, however, it is preferable to start with a prototype and stabilize it over multiple releases, while receiving feedback from users.
To prevent accidental use, such features can be only accessed by enabling the experimental mode.

There are no backwards compatibility guarantees for experimental features.
They are subject to change in breaking ways in non-breaking releases of the compiler.
Only major changes affecting them are recorded in the changelog.

To enable the experimental mode, use the ``--experimental`` flag on the command line,
or the analogous ``settings.experimental`` boolean setting in the Standard JSON input.

Note that the use of this mode is recorded in the metadata:

- ``experimental`` flag in CBOR metadata is set to ``true``,
- ``settings.experimental`` in JSON metadata is set to ``true``.
```

.. note::
    Prior to version 0.8.34, most of the experimental features were usable without any extra safeguards.
    Some were gated behind ``pragma experimental``, but this was not done consistently.
    The information about them was also only recorded in CBOR metadata and even then not always.

>     The main goal of the experimental mode is to systematize this and make users fully when they are relying on features which are unfinished or not production-ready.

Reviewer: Looks like I missed a word in my suggestion :)

```suggestion
    The main goal of the experimental mode is to systematize this and make users fully aware of when they are relying on features which are unfinished or not production-ready.
```

The table below details all currently available experimental features.

+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+
| Feature               | ID                       | Affects bytecode | Flag/pragma                                                       |
+=======================+==========================+==================+===================================================================+
| AST import            | ``ast-import``           | yes              | ``--import-ast``                                                  |
+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+

> | LSP                   | ``lsp``                  | yes              | ``--lsp``                                                         |

Reviewer: Let's change this to "no":
```suggestion
| LSP                   | ``lsp``                  | no               | ``--lsp``                                                         |
```

I put "yes" in the table in the spec defensively assuming that if any compilation was to be performed for any reason in this mode, it would be necessarily done by the server and potentially affected by any bugs in it.

But now that I think of it, there's nothing in the LSP spec that requires compilation, and even if there was, the only features we have in ours are those for which analysis is sufficient: https://github.com/argotorg/solidity/blob/81070d39868353ddada3f6c37ce052061f14ebec/libsolidity/lsp/LanguageServer.cpp#L267-L269

+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+
| EVM Assembly import   | ``evmasm-import``        | yes              | ``--import-asm-json``                                             |
+-----------------------+--------------------------+------------------+-------------------------------------------------------------------+

> | Core Solidity         | ``core-solidity``        | yes              | ``pragma experimental solidity``                                  |

Reviewer: ```suggestion
| Deprecated initial prototype of Core Solidity | ``core-solidity``        | yes              | ``pragma experimental solidity``                                  |
```

Or perhaps we should add a "remarks" column. We'll likely use the same name when we start implementing Core Solidity. The point is just to avoid people assuming that this is what Core Solidity currently looks like.

Author: I generally think the previous naming scheme fits better for this version (Generic Solidity); Calling this Core Solidity when we've invested so much in marketing it, and then saying we actually had this old thing that is now deprecated (it's technically not) and then replacing it with a different implementation (actual Core Solidity) whilst retaining the same name and ID feels very off to me.

Reviewer: ok, you have a point. TBH I would not mind having a distinct name to refer to this old prorotype. This always makes it easier to precisely refer to things. If that makes things easier from marketing perspective as well, let's go with that and revert back to `generic-solidity` here.

Reviewer: Marking as unresolved, because I still see `core-solidity` in the table. I think you changed it only in the changelog message.
